src/crimson/os/seastore/cache.cc

```diff
@@ -313,6 +313,15 @@ std::optional<record_t> Cache::try_construct_record(Transaction &t)
       });
   }
 
+  for (auto b : t.allocated_blocks) {
```

Comment:
Not sure how does this detect allocation conflicts if there are two concurrent transactions.

Reply:
Presumably there'd need to be some projected soft state with rollback if the transaction isn't submitted or conflicts?  Dunno, we'd need some more information here.

Reply:
Can you explain when two concurrent transactions happen? Those transactions are the same?
We keep the allocation information by the in-memory allocator, so can find conflicts regarding allocations.

Reply:
IIUC SeaStore only serialize transactions from the same PG, and allows to submit transactions concurrently from different PGs. Concurrent transactions may result in conflicts, for example, if they are modifing the same LBA/Onode tree node extents that happen to be shared between PGs.

OK, I see your point. Allocations won't conflict each other because they are from the same in-memory allocator, and this is why allocation-rollback and soft-free is needed.

Reply:
That combination should work, but note that currently the only notification will be upon Transaction destruction, so you'd need to hook into the Transaction destructor to make it work.  I'd probably want some kind of Cache::handle_transaction_destroy handler to make this kind of thing easy to find.

Reply:
Right. if we add Cache::handle_transaction_destroy, in-memory allocator can free unused allocation (not committed).

Reply:
This hasn't been addressed yet (not necessarily a problem at this time).